plugins/flatpak/gs-flatpak.c

```diff
@@ -1777,31 +1755,29 @@ gs_plugin_refine_item_state (GsFlatpak *self,
 				  "%s::refine-action",
 				  gs_flatpak_get_id (self));
 	g_assert (ptask != NULL);
-	xrefs = flatpak_installation_list_installed_refs (self->installation,
-							  cancellable, error);
-	if (xrefs == NULL) {
-		gs_flatpak_error_convert (error);
-		return FALSE;
-	}
-	for (i = 0; i < xrefs->len; i++) {
-		FlatpakInstalledRef *xref = g_ptr_array_index (xrefs, i);
-
-		/* check xref is app */
-		if (!gs_flatpak_app_matches_xref (self, app, FLATPAK_REF(xref)))
-			continue;
 
-		/* mark as installed */
+	ref = get_installed_ref_for_app (self->installation, app, cancellable,
+					 &local_error);
+	if (ref != NULL) {
 		g_debug ("marking %s as installed with flatpak",
 			 gs_app_get_id (app));
-		gs_flatpak_set_metadata_installed (self, app, xref);
+		gs_flatpak_set_metadata_installed (self, app, ref);
 		if (gs_app_get_state (app) == AS_APP_STATE_UNKNOWN)
 			gs_app_set_state (app, AS_APP_STATE_INSTALLED);
+	} else if (!g_error_matches (local_error, FLATPAK_ERROR, FLATPAK_ERROR_NOT_INSTALLED)) {
+		/* if there was an error that wasn't simply that the app isn't installed, then
+		 * we need to return it*/
+		g_propagate_error (error, g_steal_pointer (&local_error));
+		gs_flatpak_error_convert (error);
+		return FALSE;
 	}
 
 	/* ensure origin set */
 	if (!gs_plugin_refine_item_origin (self, app, cancellable, error))
 		return FALSE;
 
+	g_clear_error (&local_error);
```

Comment:
This should be moved up, right after you know you won't use `local_error` again (line 1774)